### PR TITLE
Allow changing namespaces for child process #6

### PR DESCRIPTION
**Pull request by anti-social**

### CODE REVIEW

File: src/child.rs [hunk 1]
```diff
@@ -195,5 +205,7 @@ mod ffi {
     extern {
         pub fn pivot_root(new_root: *const c_char, put_old: *const c_char)
             -> c_int;
+
+        pub fn setns(fd: c_int, nstype: c_int) -> c_int;
```
Comment (tailhook) [inline src/child.rs]: Anything wrong with `nix::sched::setns` ?

Comment (anti-social) [inline src/child.rs]: I think no. But I decided to do as `pivot_root` was done.

File: src/linux.rs [hunk 2]
```diff
@@ -136,6 +137,22 @@ impl Command {
...
+    /// `fd` argument is a file descriptor referring to a namespace
+    ///
+    /// 'ns' is an optional namespace type. `None` allows any type of
+    /// namespace
+    ///
+    /// See `man 2 setns` for further details
+    pub fn setns(&mut self, fd: RawFd, ns: Option<Namespace>)
+        -> &mut Command
+    {
+        self.config.setns_namespaces.push((fd, ns));
```
Comment (tailhook) [inline src/linux.rs]: There are two problems with this line:

1. We want a safe interface. It means we must ensure that file descriptor is alive as long as the command is alive. Which means two things:
 a) we must duplicate FD, or accept `IntoRawFd` object instead (which ensures that we get owned descriptor in result)
 b) we must close file descriptor after command is dropped (in parent process)

2. We usually have setter methods which replace the value instead of appending to the list. I mean if there is no sense in `x.setns(1, Namespace::Network).setns(2, Namespace::Network)`, the second call should make the first one useless. Just like `x.env("A", "B").env("A", "C")` leaves `C` in `A`. Instead of making a messy env with a duplicated environment variables. While I undrestand that `None` as `ns` ruins the whole idea of making namespaces a dictionary, I'd rather drop it to make semantics more clear. Because I don't have any actually useful use cases for `setns(x, None)` (except for lazyness)

File: src/linux.rs [hunk 3]
```diff
@@ -136,6 +137,22 @@ impl Command {
...
+    /// Reassociate child process with a namespace specified by a file
+    /// descriptor
+    ///
+    /// `fd` argument is a file descriptor referring to a namespace
+    ///
+    /// 'ns' is an optional namespace type. `None` allows any type of
+    /// namespace
+    ///
+    /// See `man 2 setns` for further details
+    pub fn setns(&mut self, fd: RawFd, ns: Option<Namespace>)
```
Comment (tailhook) [inline src/linux.rs]: Usually, we decipher the names. I.e this one should be `set_namespace` or `attach_namespace` or `associate_with_namespace`

File: src/child.rs (hunk 1 above)
Comment (anti-social) [inline src/child.rs]: `nix::sched::setns` appeared in `nix==0.5.0` and it breaks several places in `unshare`. So I think updating `nix` should be another PR.

File: src/linux.rs [hunk 4]
```diff
@@ -136,6 +137,22 @@ impl Command {
...
 
+    /// Reassociate child process with a namespace specified by a file
+    /// descriptor
+    ///
+    /// `fd` argument is a file descriptor referring to a namespace
+    ///
+    /// 'ns' is an optional namespace type. `None` allows any type of
+    /// namespace
+    ///
+    /// See `man 2 setns` for further details
```
Comment (tailhook) [inline src/linux.rs]: We should mention here, that using `unshare` and `setns` for the same namespace is meaningless.

File: src/child.rs (hunk 1 above)
Comment (tailhook) [inline src/child.rs]: Ah, sorry, we don't use `nix` in `child.rs` anyway (there is a comment about that). But `libc::setns` should be fine.

File: src/linux.rs (hunk 2 above)
Comment (anti-social) [inline src/linux.rs]: 2. Initially I kept namespaces in dictionary but then `Option` appeared.

Comment (tailhook) [inline src/linux.rs]: Well, so this looks more obvious, right?

At the end of the day, we don't need to support 100% of possible calls of `setns` just 99% of the real use cases. For now, I think dropping support for `None` is viable.

**Comment by anti-social** [comment]
Updated.

File: src/linux.rs [hunk 5]
```diff
@@ -136,6 +140,30 @@ impl Command {
...
+        -> io::Result<&mut Command>
+    {
+        let fd = match dup(file.as_raw_fd()) {
+            Ok(fd) => fd,
+            Err(nix::Error::Sys(errno)) => {
+                return Err(io::Error::from_raw_os_error(errno as i32));
+            },
+            Err(nix::Error::InvalidPath) => unreachable!(),
+        };
+        self.config.setns_namespaces.insert(ns, fd);
```
Comment (tailhook) [inline src/linux.rs]: Here you silently replace old file descriptor without closing.

File: src/lib.rs [hunk 6]
```diff
@@ -79,6 +79,14 @@ pub struct Command {
     id_map_commands: Option<(PathBuf, PathBuf)>,
 }
 
+impl Drop for Command {
+    fn drop(&mut self) {
+        for (_, fd) in &self.config.setns_namespaces {
+            unsafe { libc::close(*fd) };
```
Comment (tailhook) [inline src/lib.rs]: This should be done by making a wrapper and a Drop implementation, similarly to how is it done for passed file descriptors. This prevents the bug like in `set_namespace` that I've commented moment ago.

Comment (tailhook) [inline src/lib.rs]: To clarify: I mean a wrapper around file descriptor. Similarly to how it's done for `unshare::stdio::Fd`

**Comment by tailhook** [comment]
```
/// Note: using `unshare` and `setns` for the same namespace is meaningless.
```
This shouldn't have been written literally. I.e. `setns` basically overrides `unshare`.

Anyway I'm merging it in.
